build/common/bitbake_util.py

```diff
@@ -0,0 +1,158 @@
+
+        Mandatory args:
+        * command (str): the command to run in the environment
+
+        Optional args:
+        * timeout (int): how long to wait for the next prompt. None means to
+                         wait indefinitely
+        * stdout (bool): flag to print the output to stdout
+
+        Return:
+        * output (str): output of the specified command
+
+        """
+        output = self.shell.run_command(command, timeout=timeout, async_=False)
+        if stdout:
+            print(output)
+        return output
+
+    def run_commands(self, commands, timeout=None, stdout=True):
+        """
+        Run a list of commands in the Bitbake environment.
+
+        Mandatory args:
+        * commands (list): the commands to run in the environment
+
+        Optional args:
+        * timeout (int): how long to wait for the next prompt. None means to
+                         wait indefinitely
+        * stdout (bool): flag to print the output to stdout
+
+        If a command fails its execution, it will continue executing the
+        remaining commands.
+
+        """
+        for command in commands:
+            self.run_command(command, timeout=timeout, stdout=stdout)
+
+    def save_artifacts(self, *files):
+        """
+        Copy *files into self.outputdir directory.
+
+        Mandatory args:
+        * files (str arguments): the list specifies files/directory to be
+                                 copied into self.output directory
+
+
+        If a file/directory is not found, it will be skipped and it continues
+        to copy the next ones.
+
+        """
+        for obj in files:
+            obj_path = self.builddir / obj
+            if obj_path.exists():
+                if obj_path.is_file():
+                    copy2(obj_path, self.outputdir)
+                elif obj_path.is_dir():
+                    copytree(obj_path, self.outputdir / obj_path.name)
+                else:
+                    print(
+                        "{}: file type not recognised. Skipping.".format(
+                            obj_path
+                        ),
+                        file=sys.stderr,
+                    )
+            else:
+                print("{} doesn't exists. Skipping.".format(obj_path))
+
+    # Private methods
+    def _check_environment(self):
+        repo_dir = self.builddir / ".repo"
+        if not repo_dir.exists() or not repo_dir.is_dir():
+            raise BitbakeInvalidEnvironment(repo_dir)
+        if not self.outputdir.exists() or not self.outputdir.is_dir():
+            raise BitbakeInvalidEnvironment(self.outputdir)
+        init_env_file_path = self.builddir / self.init_env_file
+        if not init_env_file_path.exists() or not init_env_file_path.is_file():
+            raise BitbakeInvalidEnvironment(init_env_file_path)
+
+    def _build_env_variables_string(self):
+        string = ""
+        if isinstance(self.env_variables, dict):
```

Comment:
`self.env_variables` sometimes isn't a `dict`? I'd suggest we add a comment to the class docstring enumerating the cases where it isn't a `dict`.

Reply:
Addressed: d42afb6abe63d39574348df7ee963cc370f383ab